model/PermissionsService.php

```diff
@@ -54,61 +52,162 @@ public function __construct(
         $this->eventManager = $eventManager;
     }
 
-    public function saveResourcePermissionsRecursive(
-        core_kernel_classes_Resource $resource,
-        array $privilegesToSet
-    ): void {
-        $this->saveResourcePermissions($resource, $privilegesToSet, true);
-    }
-
-    private function saveResourcePermissions(
-        core_kernel_classes_Resource $resource,
-        array $privilegesToSet,
-        bool $isRecursive
-    ): void {
-        $currentPrivileges = $this->dataBaseAccess->getResourcePermissions($resource->getUri());
-        $addRemove = $this->strategy->normalizeRequest($currentPrivileges, $privilegesToSet);
+    /**
+     * Updates the permissions for a set of resources based on the ACLs, root
+     * resource and recursion parameters contained in the provided command.
+     *
+     * - For recursive commands having a class as the root, updates permissions
+     *   for that class and all its descendants (i.e. updates all resources AND
+     *   classes using the provided root class as the initial node for a tree
+     *   traversal, which may be slow and resource-intensive). This is provided
+     *   for backward compatibility purposes.
+     *
+     * - For non-recursive commands having a class as the root AND including
+     *   nested resources, updates permissions for the class and all instances
+     *   of that class, but skips all nested classes and instances of them (i.e.
+     *   does not go down into nested levels of the resource tree).
+     *
+     * - Otherwise (i.e. non-class roots), it updates only the resource set as
+     *   the root resource for the command.
+     */
```

Comment:
Suggestion: 

Maybe we should move this docs to `ChangePermissionsCommand::applyToNestedResources` and `ChangePermissionsCommand:: isRecursive`?

Reply:
Makes sense :ok_hand:

Reply:
Done: 4c4ee666484cbcb7ecf87d10ad910716506284f9